src/runner.rs

```diff
@@ -623,418 +597,401 @@ impl<'a, E: Externals> Interpreter<'a, E> {
-		let (left, right) = context
-			.value_stack_mut()
-			.pop_pair_as::<T>()
-			.expect("Due to validation stack should contain pair of values");
+		let (left, right) = self
+			.value_stack
+			.pop_pair_as::<T>();
 		let (left, right) = (left.transmute_into(), right.transmute_into());
 		let v = left.rem(right)?;
 		let v = v.transmute_into();
-		context.value_stack_mut().push(v.into())?;
+		self.value_stack.push(v.into())?;
 		Ok(InstructionOutcome::RunNextInstruction)
 	}
 
-	fn run_and<T>(&mut self, context: &mut FunctionContext) -> Result<InstructionOutcome, TrapKind>
+	fn run_and<T>(&mut self) -> Result<InstructionOutcome, TrapKind>
 	where RuntimeValue: From<<T as ops::BitAnd>::Output>, T: ops::BitAnd<T> + FromRuntimeValue {
-		let (left, right) = context
-			.value_stack_mut()
-			.pop_pair_as::<T>()
-			.expect("Due to validation stack should contain pair of values");
+		let (left, right) = self
+			.value_stack
+			.pop_pair_as::<T>();
 		let v = left.bitand(right);
-		context.value_stack_mut().push(v.into())?;
+		self.value_stack.push(v.into())?;
 		Ok(InstructionOutcome::RunNextInstruction)
 	}
 
-	fn run_or<T>(&mut self, context: &mut FunctionContext) -> Result<InstructionOutcome, TrapKind>
+	fn run_or<T>(&mut self) -> Result<InstructionOutcome, TrapKind>
 	where RuntimeValue: From<<T as ops::BitOr>::Output>, T: ops::BitOr<T> + FromRuntimeValue {
-		let (left, right) = context
-			.value_stack_mut()
-			.pop_pair_as::<T>()
-			.expect("Due to validation stack should contain pair of values");
+		let (left, right) = self
+			.value_stack
+			.pop_pair_as::<T>();
 		let v = left.bitor(right);
-		context.value_stack_mut().push(v.into())?;
+		self.value_stack.push(v.into())?;
 		Ok(InstructionOutcome::RunNextInstruction)
 	}
 
-	fn run_xor<T>(&mut self, context: &mut FunctionContext) -> Result<InstructionOutcome, TrapKind>
+	fn run_xor<T>(&mut self) -> Result<InstructionOutcome, TrapKind>
 	where RuntimeValue: From<<T as ops::BitXor>::Output>, T: ops::BitXor<T> + FromRuntimeValue {
-		let (left, right) = context
-			.value_stack_mut()
-			.pop_pair_as::<T>()
-			.expect("Due to validation stack should contain pair of values");
+		let (left, right) = self
+			.value_stack
+			.pop_pair_as::<T>();
 		let v = left.bitxor(right);
-		context.value_stack_mut().push(v.into())?;
+		self.value_stack.push(v.into())?;
 		Ok(InstructionOutcome::RunNextInstruction)
 	}
 
-	fn run_shl<T>(&mut self, context: &mut FunctionContext, mask: T) -> Result<InstructionOutcome, TrapKind>
+	fn run_shl<T>(&mut self, mask: T) -> Result<InstructionOutcome, TrapKind>
 	where RuntimeValue: From<<T as ops::Shl<T>>::Output>, T: ops::Shl<T> + ops::BitAnd<T, Output=T> + FromRuntimeValue {
-		let (left, right) = context
-			.value_stack_mut()
-			.pop_pair_as::<T>()
-			.expect("Due to validation stack should contain pair of values");
+		let (left, right) = self
+			.value_stack
+			.pop_pair_as::<T>();
```

Comment:
is ::<T> needed here ?

Reply:
In this particular, yes, it is needed. Rustc is having problems with infering the types.